### PR TITLE
Add support to add_method for classmethods, __new__, and staticmethods

### PR DESCRIPTION
**Pull request by dmontagu**
Adds keyword arguments to `add_method` to create classmethods, staticmethods, and `__new__`. Hopefully this makes it easier to add such methods in mypy plugins.

This is based on the implementation in semanal_namedtuple.py,  as suggested by @ilevkivskyi in #7301.

I still need to add tests, but I wanted to make sure this was along the right path before figuring that out.

### CODE REVIEW

**Comment by dmontagu** [comment]
@ilevkivskyi is this what you had in mind? Is there a good way for me to go about adding tests for this even though it isn't used by any of the provided plugins?

File: mypy/plugins/common.py [hunk 1]
```diff
@@ -86,6 +86,9 @@ def add_method(
         return_type: Type,
         self_type: Optional[Type] = None,
         tvar_def: Optional[TypeVarDef] = None,
```
Comment (dmontagu) [inline mypy/plugins/common.py]: Does this mean that you can currently only include a single TypeVar when using this method? Is there a good reason for that limitation?

**Comment by ilevkivskyi** [comment]
I think it would be fair if @msullivan will review this. Fro myself I can add two large scale comments:
* This definitely needs tests
* IIRC `@classmethod` and `@staticmethod` are removed from the decorator lists by the semantic analyzer, so we need to be sure to match this here (you can try inspecting generated and actual class methods).

**Comment by dmontagu** [comment]
@ilevkivskyi Thanks for taking the time to give feedback!

------

I'm using essentially this code in https://github.com/samuelcolvin/pydantic/blob/97c9b5cdbff9b83bed38ef243488eed01b070e93/pydantic/mypy.py (minus staticmethod handling) and noticed something weird that I think might be related to the point about the difference between generated and actual class methods.

Calls to a method added with `is_classmethod=True` type check properly when called via the class, but don't behave properly when called via an instance of the class. In particular, if I call `reveal_type(instance.classmethod)`, I get the following difference:

For an actual classmethod:
```
note: Revealed type is 'def (obj: Any) -> tests.mypy.plugin-success.Model*'
```

For a classmethod added via `add_method` as in this PR
```
note: Revealed type is 'def [Model] (_cls: Type[Model`-1], obj: Any) -> Model`-1'
```

Any ideas about where to look to handle this? (In particular, is this handled properly now in `semanal_namedtuple.py` and I am doing something wrong here?)

Or, perhaps related, any suggestions about what to pay attention to if I "[inspect] generated and actual class methods"?

--------

Separately, are there any tests in the codebase now that test plugin-oriented functionality? As far as I can see, for example, the `add_method` function as it exists currently is only used in the attrs and dataclasses plugins, and presumably only tested indirectly through tests of the output with those plugins enabled?

I'm happy to write tests for the added functionality, but I don't know what would be idiomatic in the mypy codebase, or even where to start looking to begin to answer that question. Should I create a "test" plugin? Is there a way to *unit*-test a function like this? Are there any reasonably analogous examples in existing tests?

**Comment by ilevkivskyi** [comment]
> Or, perhaps related, any suggestions about what to pay attention to if I "[inspect] generated and actual class methods"?

I mean have your plugin activated, then manually write a method that should like a method you want generated. The set a breakpoint somewhere in `mypy/checker.py` and in the debugger inspect both methods (e.g. corresponding `FuncDef`s and their attributes) to see that they are identical.
> Separately, are there any tests in the codebase now that test plugin-oriented functionality?

There are (relatively) many tests, see e.g. https://github.com/python/mypy/blob/master/test-data/unit/check-custom-plugin.test. The corresponding test plugins live in https://github.com/python/mypy/tree/master/test-data/unit/plugins

**Comment by dmontagu** [comment]
Thank you! This is very helpful!

**Comment by 97littleleaf11** [comment]
Hi @dmontagu , any updates on this PR?

**Pull request closed without merging**